data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,163 @@
+from pyspark.sql.types import IntegerType, StringType
+
+
+BUCKET_NAME = sys.argv[1]
+TABLE = "bigquery-public-data.new_york_citibike.citibike_trips"
+
+
+# START MAKING DATA DIRTY
+def random_select(items, weights):
+    '''Picks an item according to the cumulative weights'''
+    return random.choices(items, weights=weights, k=1)[0]
+
+
+def trip_duration(duration):
+    '''Converts trip duration to other units'''
+    seconds = str(duration) + " s"
+    minutes = str(float(duration) / 60) + " min"
+    hours = str(float(duration) / 3600) + " h"
+    return random_select([seconds, minutes, hours,
+                         str(random.randint(-1000, -1))],
+                         [0.3, 0.3, 0.3, 0.1])
+
+
+def station_name(name):
+    '''Replaces '&' with '/' with a 50% chance'''
+    return random.choice([name, name.replace("&", "/")])
+
+
+def user_type(user):
+    '''Manipulates the user type string'''
+    return random.choice([user, user.upper(), user.lower(),
+                          "sub" if user == "Subscriber" else user,
+                          "cust" if user == "Customer" else user])
+
+
+def gender(s):
+    '''Manipulates the gender string'''
+    return random.choice([s.upper(), s.lower(),
+                         s[0].upper() if len(s) > 0 else "",
+                         s[0].lower() if len(s) > 0 else ""])
+
+
+def convert_angle(angle):
+    '''Converts long and lat to DMS notation'''
+    degrees = int(angle)
+    minutes = int((angle - degrees) * 60)
+    seconds = int((angle - degrees - minutes/60) * 3600)
+    new_angle = str(degrees) + u"\u00B0" + \
+        str(minutes) + "'" + str(seconds) + '"'
+    return random_select([str(angle), new_angle], [0.55, 0.45])
+
+
+# This function is nested since a UserDefinedFunction is
+# expected to take a single argument
+def dirty_data(proc_func, allow_none):
+    '''Master function returns a user defined function
+    that transforms the column data'''
+    def udf(col_value):
+        random.seed(hash(col_value) + time_ns())
+        if col_value is None:
+            return col_value
+        elif allow_none:
+            return random_select([None, proc_func(col_value)],
+                                 [0.05, 0.95])
+        else:
+            return proc_func(col_value)
+    return udf
+
+
+# This function is required because we need to apply a
+# function for every column and some columns do not change
+def identity(x):
+    return x
+
+
+def write_to_bigquery(spark, df):
+    '''Write a dataframe to BigQuery'''
+
+    # Create BigQuery Dataset
+    client = bigquery.Client()
+    dataset_id = f'{client.project}.new_york_citibike_trips'
+    dataset = bigquery.Dataset(dataset_id)
+    dataset.location = "US"
+    dataset = client.create_dataset(dataset)
+
+    # Saving the data to BigQuery
+    spark.conf.set('temporaryGcsBucket', BUCKET_NAME)
+
+    df.write.format('bigquery') \
+        .option('table', dataset_id + ".RAW_DATA") \
```

Comment:
nit: you can move the temporary bucket into the `write` operation to avoid editing the Spark conf:
```
df.write
    .format("bigquery")
    .option("table", dataset_id + ".RAW_DATA")
    .option("temporaryGcsBucket", BUCKET_NAME)
    .save()
```